server/service/src/plugin_files.rs

```diff
@@ -0,0 +1,90 @@
+use std::{fs, io::Error, path::PathBuf, str::FromStr};
+
+#[derive(Debug, PartialEq)]
+pub struct PluginFile {
+    pub config: String,
+    pub name: String,
+    pub path: String,
+}
+
+const PLUGIN_FILE_DIR: &'static str = "plugins";
+
+pub struct PluginFileService {
+    pub dir: PathBuf,
+}
+
+impl PluginFileService {
+    pub fn new(base_dir: &Option<String>) -> anyhow::Result<Self> {
+        let file_dir = match base_dir {
+            Some(file_dir) => PathBuf::from_str(file_dir)?.join(PLUGIN_FILE_DIR),
+            None => PathBuf::from_str(PLUGIN_FILE_DIR)?,
+        };
+
+        Ok(PluginFileService { dir: file_dir })
+    }
+
+    pub fn find_file(
+        &self,
+        plugin: &str,
+        filename: Option<&str>,
+    ) -> anyhow::Result<Option<PluginFile>> {
+        fs::create_dir_all(&self.dir).unwrap();
+
+        let filename = match filename {
+            Some(filename) => filename.to_string(),
+            None => format!("{}.js", plugin).clone(),
+        };
+        let file_path = match find_file(&self.dir, plugin, &filename)? {
+            Some(path) => path,
+            None => return Ok(None),
+        };
+
+        Ok(Some(PluginFile {
+            name: plugin.to_string(),
+            path: file_path.to_string_lossy().to_string(),
+            config: fs::read_to_string(file_path.parent().unwrap().join("plugin.json"))?,
+        }))
+    }
+
+    pub fn find_files(&self) -> anyhow::Result<Vec<PluginFile>> {
```

Comment:
my intention here is to shift this to the server startup - which could enumerate plugins, verify signatures and then upsert the details to the database.

we could then query off the database ( and potentially serve the files from the db too? ) and then allow disabling of plugins by the user

Reply:
yes, they could stay in memory. My only reason for using the database was to retain configuration - the intention is that plugins can be dis/en abled by users, and that the state of a plugin would be retained when the server is restarted.

the plugins themselves could live in memory though - it's just the list and the state that we need to retain